pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -79,7 +79,7 @@ def can_prepare_data(self):
         if self.trainer.datamodule is not None and is_overridden("prepare_data", self.trainer.datamodule):
             should_call_dm_prepare_data = not self.trainer.datamodule._has_prepared_data
 
-        if self.trainer.prepare_data_per_node:
+        if self.prepare_data_per_node:
```

Comment:
shouldn't this check `self.trainer.datamodule.prepare_data_per_node` ?

Reply:
Hi, @ananthsub  Good suggestion.  I think `self.trainer.datamodule.prepare_data_per_node' should only be valid when `datamodule` is not None. While in the initiation of `data_connector`, `self.trainer.datamodule` is set to be None as default in `on_trainer_init`.

Reply:
I can add `datamodule is not None` checking.

Reply:
you can see `prepare_data` above: it checks if self.trainer.datamodule is not None. This will be called only after one of fit/validate/test/predict is called with the model and datamodule. 

the risk is that `data_connector.prepare_data_per_node` is set in the trainer constructor, but we don't respect the property on the module

you can see https://github.com/PyTorchLightning/pytorch-lightning/pull/7323/files for a similar PR that deprecated a flag off the trainer in favor of adding the property to the module